File: protocol-generator/cpp/test/smoke.cpp
// smoke.cpp — the S3 phase exit gate: a two-C++-peer loopback over real localhost TCP.
//
// A RESPONDER peer listens; an INITIATOR peer (a second identity) dials it and drives the
// §4.1 forward handshake (hello → authenticate), then exercises the wire-level peer surface:
//   Scenario 1 (core ops, default seed policy):
//     - session established (capability minted)         §4.1
//     - remote peer_id matches the responder            §4.6 identity binding
//     - unregistered path → 404                          §6.6 no handler resolved
//     - granted tree get → 200 (discovery floor)         §4.4
//     - tree get returns a system/handler/interface
//     - capability request → 200                          §6.2 mint-bounded
//     - 8 interleaved requests each correlated → 8/8     N7 / §6.11 request_id demux
//   Scenario 2 (Core Extensibility Boundary; --debug-open-grants + --validate):
//     - handler register → 200 (live, not 501)           §6.13(a)
//     - emit hook fired on register's tree writes        §6.13(c)
//     - §7a echo → 200                                    §7a resolve→dispatch
//     - §7a echo returns params verbatim
//     - §7a dispatch-outbound reentry → 200 (§6.11)       the S4 surface, smoke-tested
//   → SMOKE: PASS (12/12)
//
// Built under ASan/LSan/UBSan: a memory bug (leak / UAF / overflow / UB) FAILS the run.
// The full validate-peer --profile core run is S4; this proves the peer talks the wire.
//
// SPDX-License-Identifier: Apache-2.0
#include <atomic>
#include <cstdio>
#include <thread>
#include <vector>

#include "entity_core/peer.hpp"
#include "entity_core/transport.hpp"
#include "entity_core/wire.hpp"

using namespace entity_core;

namespace {
int g_pass = 0;
int g_fail = 0;
std::atomic<int> g_emit_events{0};

void check(const char* name, bool ok) {
    if (ok) g_pass++; else g_fail++;
    std::printf("  [%s] %s\n", ok ? "PASS" : "FAIL", name);
}

std::vector<std::byte> seed_fill(std::uint8_t b) {
    return std::vector<std::byte>(32, std::byte{b});
}

EcfValue scope_inc(std::span<const std::string_view> items) {
    auto m = EcfValue::map();
    m.put(EcfValue::text("include"), value::text_array(items));
    return m;
}

EcfValue make_scoped_grant(std::span<const std::string_view> h,
                           std::span<const std::string_view> r,
                           std::span<const std::string_view> o) {
    auto g = EcfValue::map();
    g.put(EcfValue::text("handlers"), scope_inc(h));
    g.put(EcfValue::text("resources"), scope_inc(r));
    g.put(EcfValue::text("operations"), scope_inc(o));
    return g;
}

// ── scenario 1 ────────────────────────────────────────────────────────────────────
int scenario_core() {
    auto responder = Peer::create(seed_fill(0x11), false, false);
    auto initiator = Peer::create(seed_fill(0x22), false, false);
    if (!responder || !initiator) { std::fprintf(stderr, "peer create failed\n"); return 1; }

    auto listener = Listener::start(**responder, 0);
    if (!listener) { std::fprintf(stderr, "listen failed\n"); return 1; }
    int port = (*listener)->port();
    std::printf("Responder listening on 127.0.0.1:%d (peer %s)\n", port,
                (*responder)->local().c_str());

    auto sess = Session::dial(**initiator, "127.0.0.1", port);
    if (!sess) { std::fprintf(stderr, "dial/handshake failed\n"); return 1; }
    Session& s = **sess;
    const std::string& remote = s.remote_peer();

    std::printf("Handshake:\n");
    check("session established (capability minted)", s.has_capability());
    check("remote peer_id matches responder", remote == (*responder)->local());

    std::printf("Dispatch:\n");
    {
        std::string uri = "/" + remote + "/does/not/exist";
        auto params = wire::empty_params();
        auto r = s.execute(uri, "noop", **params, std::nullopt);
        check("unregistered path -> 404", r && response_status(*r) == 404);
    }
    {
        std::string uri = "/" + remote + "/system/tree";
        auto params = wire::empty_params();
        auto res = wire::resource_target("system/handler/system/tree");
        auto r = s.execute(uri, "get", **params, res);
        check("granted tree get -> 200", r && response_status(*r) == 200);
        auto rr = r ? response_result(*r) : nullptr;
        check("tree get returns a system/handler/interface entity",
              rr && rr->type() == "system/handler/interface");
    }
    {
        std::string uri = "/" + remote + "/system/capability";
        using V = std::string_view;
        std::array<V, 1> h{"system/tree"};
        std::array<V, 1> r_{"system/type/*"};
        std::array<V, 1> o{"get"};
        auto grant = make_scoped_grant(h, r_, o);
        auto grants = EcfValue::array();
        grants.push(std::move(grant));
        auto pm = EcfValue::map();
        pm.put(EcfValue::text("grants"), std::move(grants));
        auto params = Entity::make("system/capability/request", std::move(pm));
        auto r = s.execute(uri, "request", **params, std::nullopt);
        check("capability request -> 200", r && response_status(*r) == 200);
    }
    std::printf("Concurrency (request_id demux):\n");
    {
        constexpr int N = 8;
        std::array<bool, N> okv{};
        std::vector<std::thread> ts;
        std::string uri = "/" + remote + "/system/tree";
        for (int i = 0; i < N; ++i) {
            ts.emplace_back([&, i] {
                auto params = wire::empty_params();
                auto res = wire::resource_target("system/handler/system/tree");
                auto r = s.execute(uri, "get", **params, res);
                auto rr = r ? response_result(*r) : nullptr;
                okv[i] = r && response_status(*r) == 200 && rr &&
                         rr->type() == "system/handler/interface";
            });
        }
        int correlated = 0;
        for (int i = 0; i < N; ++i) { ts[i].join(); if (okv[i]) correlated++; }
        char msg[64];
        std::snprintf(msg, sizeof(msg), "8 interleaved requests each correlated -> %d/8",
                      correlated);
        check(msg, correlated == N);
    }
    return 0;
}

// ── scenario 2 ────────────────────────────────────────────────────────────────────
int scenario_extensibility() {
    auto responder = Peer::create(seed_fill(0x33), true, true);   // open-grants + validate
    auto initiator = Peer::create(seed_fill(0x44), false, false);
    if (!responder || !initiator) { std::fprintf(stderr, "peer create failed\n"); return 1; }
    (*responder)->store().register_tree_consumer(
        [](const std::string&) { g_emit_events.fetch_add(1); });

    auto listener = Listener::start(**responder, 0);
    if (!listener) return 1;
    int port = (*listener)->port();
    auto sess = Session::dial(**initiator, "127.0.0.1", port);
    if (!sess) return 1;
    Session& s = **sess;
    const std::string& remote = s.remote_peer();
    int emit_before = g_emit_events.load();

    std::printf("Extensibility (open-grants + --validate):\n");
    {
        std::string uri = "/" + remote + "/system/handler";
        auto manifest = EcfValue::map();
        manifest.put(EcfValue::text("name"), EcfValue::text("demo"));
        manifest.put(EcfValue::text("operations"), EcfValue::map());
        auto pm = EcfValue::map();
        pm.put(EcfValue::text("manifest"), std::move(manifest));
        auto req = Entity::make("system/handler/register-request", std::move(pm));
        auto res = wire::resource_target("system/handler/demo");
        auto r = s.execute(uri, "register", **req, res);
        check("handler register -> 200 (live, not 501)", r && response_status(*r) == 200);
        check("emit hook fired on register's tree writes (§6.13(c))",
              g_emit_events.load() > emit_before);
    }
    {
        std::string uri = "/" + remote + "/system/validate/echo";
        auto pm = EcfValue::map();
        pm.put(EcfValue::text("ping"), EcfValue::uint(42));
        auto payload = Entity::make("primitive/any", std::move(pm));
        auto r = s.execute(uri, "echo", **payload, std::nullopt);
        check("§7a echo -> 200", r && response_status(*r) == 200);
        auto rr = r ? response_result(*r) : nullptr;
        check("§7a echo returns params verbatim", rr && rr->type() == "primitive/any");
    }
    return 0;
}

// ── scenario 3: §6.11 reentry dispatch-outbound (the S4 origination-core surface) ──
//
// The INITIATOR is also a B-role server on its own connection: it minted a reentry cap
// authorizing the RESPONDER to call the initiator's system/validate/echo. The initiator
// sends `system/validate/dispatch-outbound:dispatch` to the responder; the responder
// ORIGINATES an EXECUTE back to the initiator over the SAME inbound connection (§6.11
// reentry), the initiator's reader dispatches it (validator-as-B), and the echo result
// flows back. This proves the reentry transport + correlation the S4 gate (3/3) needs.
int scenario_reentry() {
    // Both peers conformance-enabled: the responder runs dispatch-outbound; the initiator
    // serves echo as the reentry B-role.
    auto responder = Peer::create(seed_fill(0x55), true, true);
    auto initiator = Peer::create(seed_fill(0x66), true, true);
    if (!responder || !initiator) { std::fprintf(stderr, "peer create failed\n"); return 1; }

    auto listener = Listener::start(**responder, 0);
    if (!listener) return 1;
    int port = (*listener)->port();
    auto sess = Session::dial(**initiator, "127.0.0.1", port);
    if (!sess) return 1;
    Session& s = **sess;
    const std::string& remote = s.remote_peer();
    const std::string& local_init = (*initiator)->local();
    const auto& init_id = (*initiator)->identity();
    const auto& resp_id = (*responder)->identity();

    std::printf("Reentry (§6.11 dispatch-outbound over the same connection):\n");

    // Mint a reentry cap: granter = initiator, grantee = responder, granting echo on the
    // initiator's namespace. Signed by the initiator (validator-as-B authority).
    using V = std::string_view;
    auto scope = [](std::span<const V> items) {
        auto m = EcfValue::map();
        m.put(EcfValue::text("include"), value::text_array(items));
        return m;
    };
    std::array<V, 1> h{"system/validate/echo"};
    std::array<V, 1> o{"echo"};
    std::array<V, 1> star{"*"};
    std::array<V, 1> peers{local_init};
    auto grant = EcfValue::map();
    grant.put(EcfValue::text("handlers"), scope(h));
    grant.put(EcfValue::text("resources"), scope(star));
    grant.put(EcfValue::text("operations"), scope(o));
    grant.put(EcfValue::text("peers"), scope(std::span<const V>(peers)));
    auto grants = EcfValue::array();
    grants.push(std::move(grant));
    auto tm = EcfValue::map();
    tm.put(EcfValue::text("granter"), value::bytes_value(init_id.identity_hash()));
    tm.put(EcfValue::text("grantee"), value::bytes_value(resp_id.identity_hash()));
    tm.put(EcfValue::text("grants"), std::move(grants));
    tm.put(EcfValue::text("created_at"), EcfValue::uint(cap::now_ms()));
    auto reentry_cap = Entity::make("system/capability/token", std::move(tm));
    auto reentry_sig = init_id.sign(**reentry_cap);
    auto reentry_granter = init_id.peer_entity();   // the granter's system/peer entity

    // The dispatch-outbound params: {target, operation, value, reentry_*}. The value is the
    // echo payload, forwarded verbatim.
    auto value_map = EcfValue::map();
    value_map.put(EcfValue::text("marker"), EcfValue::text("reentry-roundtrip"));
    auto pm = EcfValue::map();
    pm.put(EcfValue::text("target"), EcfValue::text("system/validate/echo"));
    pm.put(EcfValue::text("operation"), EcfValue::text("echo"));
    pm.put(EcfValue::text("value"), std::move(value_map));
    pm.put(EcfValue::text("reentry_capability"), (*reentry_cap)->to_cbor());
    pm.put(EcfValue::text("reentry_granter"), reentry_granter->to_cbor());
    pm.put(EcfValue::text("reentry_cap_signature"), (*reentry_sig)->to_cbor());
    auto params = Entity::make("primitive/any", std::move(pm));

    std::string uri = "/" + remote + "/system/validate/dispatch-outbound";
    auto res = wire::resource_target("system/handler/system/validate/dispatch-outbound");
    auto r = s.execute(uri, "dispatch", **params, res);
    bool ok200 = r && response_status(*r) == 200;
    check("dispatch-outbound -> 200 (reentry produced a response)", ok200);
    // the result wraps the downstream {status, result}; the downstream echo status is 200.
    if (r) {
        auto rr = response_result(*r);
        auto dstatus = rr ? rr->uint("status") : std::nullopt;
        check("downstream echo via reentry -> status 200",
              dstatus.has_value() && *dstatus == 200);
    } else {
        check("downstream echo via reentry -> status 200", false);
    }
    return 0;
}

}  // namespace

int main() {
    if (auto i = crypto::init(); !i) { std::fprintf(stderr, "crypto init failed\n"); return 1; }
    if (scenario_core() != 0) { std::printf("\nSMOKE: FAIL (harness error in scenario 1)\n"); return 1; }
    if (scenario_extensibility() != 0) { std::printf("\nSMOKE: FAIL (harness error in scenario 2)\n"); return 1; }
    if (scenario_reentry() != 0) { std::printf("\nSMOKE: FAIL (harness error in scenario 3)\n"); return 1; }
    bool all_pass = (g_fail == 0);
    std::printf("\nSMOKE: %s (%d/%d)\n", all_pass ? "PASS" : "FAIL", g_pass, g_pass + g_fail);
    return all_pass ? 0 : 1;
}

File: protocol-generator/haskell/test/Smoke.hs
{-# LANGUAGE OverloadedStrings #-}

-- | S3 smoke runner — boots TWO Haskell peers over real loopback TCP and drives
-- the lifecycle scenario, all green:
--
--   1. §4.1 handshake (hello → authenticate; session + §4.4/§6.9a initial cap).
--   2. EXECUTE on an unregistered path → 404.
--   3. Authority-gated tree get → 200 (discovery-floor grant admits system/type/*).
--   4. capability request → 200 (mints a bounded child cap).
--   5. request_id demux (N7): N concurrent EXECUTEs each correlate to their reply.
--   6. Clean teardown (no hangs; STM/green-thread cleanup).
--   7. (--validate surface) register → dispatch round-trip + dispatch-outbound
--      reentry self-check (the §7a handlers, B-role on the same connection).
--
-- Green = the peer talks the wire correctly. The client side is built from the
-- library's own builders + codec (no fakes): it dials, runs its own dispatcher on
-- the bidirectional connection (so the §6.11 reentry leg works), signs as a real
-- second identity, and asserts statuses by request_id correlation.
module Main (main) where

import Control.Concurrent (forkIO, threadDelay)
import Control.Concurrent.MVar
import Control.Monad (forM, forM_, unless)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BC
import Data.IORef
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time.Clock.POSIX (getPOSIXTime)
import qualified Data.Word as Word
import Network.Socket
import System.Exit (exitFailure)
import System.IO (BufferMode (LineBuffering), hFlush, hSetBuffering, stdout)

import EntityCore.Codec.Value (Value (..))
import EntityCore.Identity (Identity (..), identityOfSeed, signEntity)
import EntityCore.Model
import EntityCore.Peer (createPeer)
import EntityCore.SeedPolicy (SeedPolicy (..))
import EntityCore.Transport (ConnIOHandle, acceptLoop, listenOn, runWithConn, sendOver)
import EntityCore.Wire (makeExecute)

-- ── test harness ──────────────────────────────────────────────────────────────

assertEq :: (Eq a, Show a) => Text -> a -> a -> IO Bool
assertEq label expected got
  | expected == got = do BC.putStrLn (BC.pack ("  ok   " ++ T.unpack label)); pure True
  | otherwise = do
      BC.putStrLn (BC.pack ("  FAIL " ++ T.unpack label ++ " — expected " ++ show expected ++ ", got " ++ show got))
      pure False

statusOf :: Maybe Envelope -> Int
statusOf Nothing = -1
statusOf (Just env) = maybe (-1) fromIntegral (uintField (envRoot env) "status")

-- ── client identity + envelope construction ──────────────────────────────────

data Client = Client
  { clIdent :: Identity
  , clHandle :: ConnIOHandle
  , clReqCtr :: IORef Int
  , clCap :: IORef (Maybe Entity) -- the §4.4 initial cap (the granted token)
  , clCapSig :: IORef (Maybe Entity) -- its signature
  , clGranter :: IORef (Maybe Entity) -- the server peer entity (granter)
  }

-- | Atomic request-id allocation: 'atomicModifyIORef'' so 16 concurrent senders
-- never collide on a request_id (a duplicate id would cross-wire the §6.11 demux
-- and hang a 'takeMVar' — the N7 trap, in the client this time).
nextReqId :: Client -> IO Text
nextReqId cl = do
  n <- atomicModifyIORef' (clReqCtr cl) (\x -> (x + 1, x + 1))
  pure ("req-" <> T.pack (show n))

-- | Send a self-authored EXECUTE (author + capability + signatures in included),
-- await its EXECUTE_RESPONSE by request_id.
sendAuthed :: Client -> Text -> Text -> Entity -> Maybe Value -> IO (Maybe Envelope)
sendAuthed cl uri operation params resource = do
  rid <- nextReqId cl
  mcap <- readIORef (clCap cl)
  mcapSig <- readIORef (clCapSig cl)
  mgranter <- readIORef (clGranter cl)
  let ident = clIdent cl
      capHash = maybe BS.empty entHash mcap
      exec = makeExecute rid uri operation params resource (idIdentityHash ident) capHash
      execSig = signEntity ident exec
      included =
        [ (idIdentityHash ident, idPeerEntity ident)
        , (entHash execSig, execSig)
        ]
          ++ maybe [] (\cap -> [(entHash cap, cap)]) mcap
          ++ maybe [] (\s -> [(entHash s, s)]) mcapSig
          ++ maybe [] (\g -> [(entHash g, g)]) mgranter
  sendOver (clHandle cl) (Envelope exec included)

-- ── handshake (§4.1 / §4.6) ───────────────────────────────────────────────────

handshake :: Client -> IO Bool
handshake cl = do
  let ident = clIdent cl
  -- hello (unauthenticated EXECUTE on system/protocol/connect)
  ridH <- nextReqId cl
  let helloParams = makeEntity "primitive/any" (VMap [(VText "peer_id", VText (idPeerId ident))])
      helloExec = makeExecute ridH "system/protocol/connect" "hello" helloParams Nothing (idIdentityHash ident) BS.empty
  helloResp <- sendOver (clHandle cl) (Envelope helloExec [(idIdentityHash ident, idPeerEntity ident)])
  okHello <- assertEq "handshake: hello → 200" 200 (statusOf helloResp)
  let nonce = case helloResp >>= resultEntityOf >>= (`bytesField` "nonce") of
        Just nn -> nn
        Nothing -> BS.empty
      serverPeerId = fromMaybe "" (helloResp >>= resultEntityOf >>= (`textField` "peer_id"))
  okPid <- assertEq "handshake: remote peer_id present" True (not (T.null serverPeerId))
  -- authenticate: params carry public_key + echoed nonce + claimed peer_id; the
  -- auth entity is signed (the §4.6 proof-of-possession).
  ridA <- nextReqId cl
  let authEntity =
        makeEntity
          "primitive/any"
          ( VMap
              [ (VText "public_key", VBytes (idPublicKey ident))
              , (VText "nonce", VBytes nonce)
              , (VText "peer_id", VText (idPeerId ident))
              ]
          )
      authSig = signEntity ident authEntity
      authExec = makeExecute ridA "system/protocol/connect" "authenticate" authEntity Nothing (idIdentityHash ident) BS.empty
      authIncluded =
        [ (idIdentityHash ident, idPeerEntity ident)
        , (entHash authSig, authSig)
        ]
  authResp <- sendOver (clHandle cl) (Envelope authExec authIncluded)
  okAuth <- assertEq "handshake: authenticate → 200" 200 (statusOf authResp)
  -- capture the granted cap + its signature + the granter (server) peer from the
  -- authenticate response's included set, for subsequent authed EXECUTEs.
  case authResp of
    Just env -> do
      let inc = envIncluded env
          tokenHash = resultEntityOf env >>= (`bytesField` "token")
          token = tokenHash >>= (`lookup` inc)
          tokSig = token >>= \t -> findSigFor (entHash t) inc
          granter = case [e | (_, e) <- inc, entType e == "system/peer"] of (e : _) -> Just e; _ -> Nothing
      writeIORef (clCap cl) token
      writeIORef (clCapSig cl) tokSig
      writeIORef (clGranter cl) granter
    Nothing -> pure ()
  pure (okHello && okPid && okAuth)
  where
    findSigFor h inc = case [e | (_, e) <- inc, entType e == "system/signature", bytesField e "target" == Just h] of
      (e : _) -> Just e
      _ -> Nothing

resultEntityOf :: Envelope -> Maybe Entity
resultEntityOf env = case field (envRoot env) "result" of
  Just v -> either (const Nothing) Just (entityOfCbor v)
  Nothing -> Nothing

-- ── the scenario ──────────────────────────────────────────────────────────────

main :: IO ()
main = withSocketsDo $ do
  hSetBuffering stdout LineBuffering
  BC.putStrLn "== entity-core-protocol-haskell S3 smoke =="
  -- Server peer (validate on, so the §7a surface is reachable).
  serverIdent <- either error pure (identityOfSeed (BC.replicate 32 '\x11'))
  -- Boot with the degenerate open seed policy (= --debug-open-grants) so an
  -- authenticated peer receives a wide cap — the smoke runner drives the
  -- grant-gated register + §7a dispatch-outbound surface end-to-end (exactly what
  -- the validate-peer harness does). The restricted §4.4 floor is exercised by
  -- the get/request steps above with the default policy in the dedicated tests.
  server <- createPeer (idSeed serverIdent) SeedPolicyDebugOpen True
  (lsock, port) <- listenOn 0
  _ <- forkIO (acceptLoop server lsock)
  threadDelay 100000

  -- Client peer identity (a distinct second peer).
  clientIdent <- either error pure (identityOfSeed (BC.replicate 32 '\x22'))
  -- The client also runs a Peer so its dispatcher can answer the reentrant
  -- dispatch-outbound leg (B-role on the same connection, §7a.2a). --validate so
  -- the client's own system/validate/echo is reachable for the reentry target.
  clientPeer <- createPeer (idSeed clientIdent) SeedPolicyStandard True

  -- Dial the server; get a bidirectional connection (the client's Peer serves
  -- reentrant EXECUTEs on it).
  addr : _ <- getAddrInfo (Just defaultHints {addrSocketType = Stream}) (Just "127.0.0.1") (Just (show port))
  csock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
  connect csock (addrAddress addr)
  (_clientConn, handle) <- runWithConn clientPeer csock

  ctr <- newIORef 0
  capRef <- newIORef Nothing
  capSigRef <- newIORef Nothing
  granterRef <- newIORef Nothing
  let cl = Client clientIdent handle ctr capRef capSigRef granterRef

  results <- newIORef []
  let record b = modifyIORef' results (b :)

  -- 1. handshake
  handshake cl >>= record

  -- 2. EXECUTE on an unregistered path → 404
  let bogusParams = makeEntity "primitive/any" (VMap [])
      bogusRes = VMap [(VText "targets", VArray [VText "system/nonexistent/thing"])]
  r404 <- sendAuthed cl "system/nonexistent/thing" "get" bogusParams (Just bogusRes)
  assertEq "unregistered path → 404" 404 (statusOf r404) >>= record

  -- 3. authority-gated tree get → 200 (discovery floor admits system/type/*)
  let getParams = makeEntity "primitive/any" (VMap [])
      typeRes = VMap [(VText "targets", VArray [VText "system/type/primitive/any"])]
  rGet <- sendAuthed cl "system/tree" "get" getParams (Just typeRes)
  assertEq "authority-gated tree get → 200" 200 (statusOf rGet) >>= record

  -- 4. capability request → 200 (mints a bounded child cap, subset of ours)
  let reqGrants =
        VArray
          [ VMap
              [ (VText "handlers", VMap [(VText "include", VArray [VText "system/tree"])])
              , (VText "resources", VMap [(VText "include", VArray [VText "system/type/*"])])
              , (VText "operations", VMap [(VText "include", VArray [VText "get"])])
              ]
          ]
      capParams = makeEntity "primitive/any" (VMap [(VText "grants", reqGrants)])
  -- No resource target: capability:request is addressed by handler+operation, and
  -- the §4.4 floor grant for system/capability carries no resource scope, so a
  -- resource target would fail the §5.4 resource check (the floor admits the op,
  -- not a resource).
  rCap <- sendAuthed cl "system/capability" "request" capParams Nothing
  assertEq "capability request → 200" 200 (statusOf rCap) >>= record

  -- 5. request_id demux (N7): N concurrent EXECUTEs, each correlates to its reply
  let n = 16
  done <- newEmptyMVar
  forM_ [1 .. n] $ \_ -> forkIO $ do
    rr <- sendAuthed cl "system/tree" "get" getParams (Just typeRes)
    putMVar done (statusOf rr)
  statuses <- forM [1 .. n] $ \_ -> takeMVar done
  assertEq "request_id demux: N concurrent EXECUTEs all 200" (replicate n 200) statuses >>= record

  -- 7. §7a register → dispatch round-trip (the register *contract*: five writes)
  let regManifest = VMap [(VText "name", VText "smoke-handler"), (VText "operations", VMap [(VText "ping", VMap [])])]
      regReq = makeEntity "system/handler/register-request" (VMap [(VText "manifest", regManifest)])
      regRes = VMap [(VText "targets", VArray [VText "system/handler/smoke/test"])]
  rReg <- sendAuthed cl "system/handler" "register" regReq (Just regRes)
  assertEq "register → 200 (five writes)" 200 (statusOf rReg) >>= record

  -- 7b. dispatch-outbound reentry self-check (§7a.2a): the server originates an
  -- echo back to us (B-role) over the same connection. We pass our own cap as the
  -- reentry authority (the cap valid at the caller = us).
  mcap <- readIORef capRef
  mcapSig <- readIORef capSigRef
  case (mcap, mcapSig) of
    (Just _cap, Just _capSig) -> do
      -- Mint the REENTRY capability: granted by the CLIENT (us) to the SERVER, so
      -- the server is authorized to originate back to us over the connection
      -- (§7a.2a — the reentry direction can only be authorized by the caller). A
      -- wide-open grant over our namespace; signed by the client identity. This is
      -- the cap valid AT THE CALLER (us) — distinct from the server-granted cap we
      -- use for our own outbound EXECUTEs.
      now <- fmap (round . (* 1000)) getPOSIXTime
      let openGrant =
            VMap
              [ (VText "handlers", VMap [(VText "include", VArray [VText "*"])])
              , (VText "resources", VMap [(VText "include", VArray [VText "*", VText "/*/*"])])
              , (VText "operations", VMap [(VText "include", VArray [VText "*"])])
              , (VText "peers", VMap [(VText "include", VArray [VText "*"])])
              ]
          reentryCap =
            makeEntity
              "system/capability/token"
              ( VMap
                  [ (VText "granter", VBytes (idIdentityHash clientIdent))
                  , (VText "grantee", VBytes (idIdentityHash serverIdent))
                  , (VText "grants", VArray [openGrant])
                  , (VText "created_at", VUInt (now :: Word.Word64))
                  ]
              )
          reentryCapSig = signEntity clientIdent reentryCap
          echoInner = VText "reentry-ping"
          echoVal = VMap [(VText "value", echoInner)]
          dispParams =
            makeEntity
              "primitive/any"
              ( VMap
                  [ (VText "target", VText "system/validate/echo")
                  , (VText "operation", VText "echo")
                  , (VText "value", echoVal)
                  , (VText "reentry_capability", entityToCbor reentryCap)
                  , (VText "reentry_granter", entityToCbor (idPeerEntity clientIdent))
                  , (VText "reentry_cap_signature", entityToCbor reentryCapSig)
                  ]
              )
          dispRes = VMap [(VText "targets", VArray [VText "system/validate/dispatch-outbound"])]
      rDisp <- sendAuthed cl "system/validate/dispatch-outbound" "dispatch" dispParams (Just dispRes)
      okDispStatus <- assertEq "dispatch-outbound reentry → 200" 200 (statusOf rDisp)
      -- the relayed echo result.value must equal what we sent (value pass-through)
      let echoedBack = do
            env <- rDisp
            res <- resultEntityOf env
            innerResultV <- field res "result"
            inner <- either (const Nothing) Just (entityOfCbor innerResultV)
            field inner "value"
      okDispVal <- assertEq "dispatch-outbound reentry: echo value passthrough" (Just echoInner) echoedBack
      record (okDispStatus && okDispVal)
    _ -> do
      BC.putStrLn "  FAIL dispatch-outbound: no cap captured from authenticate"
      record False

  -- 6. teardown — close the client connection; the listener + green threads are
  -- reaped on process exit (closing lsock under a blocked accept races a
  -- threadWait, so we leave it to RTS shutdown).
  close csock
  threadDelay 50000

  rs <- readIORef results
  let passed = length (filter id rs)
      total = length rs
  BC.putStrLn (BC.pack ("== smoke: " ++ show passed ++ "/" ++ show total ++ " green =="))
  hFlush stdout
  unless (and rs) exitFailure
